src/dune_manager/messages_intf.ml

```diff
@@ -0,0 +1,26 @@
+open Stdune
+
+type version =
+  { major : int
+  ; minor : int
+  }
+
+type promotion =
+  { key : Digest.t
+  ; files : (Path.Build.t * Digest.t) list
+  ; metadata : Sexp.t list
+  ; repository : int option
+  }
+
+type outgoing = Outgoing
+
+type incoming = Incoming
+
+type _ message =
```

Comment:
Why use a GADT rather than two separate types?

Reply:
To have some functions that can work on messages regardless of the direction (e.g. `val send : _ message -> out_channel -> unit`)